### PR TITLE
Keysend implementation

### PR DESCRIPTION
**Pull request by skosito**

### CODE REVIEW

File: controllers/keysend.ctrl.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
+package controllers
```
Comment (skosito) [inline controllers/keysend.ctrl.go:1]: @kiwiidb compared to your changes, I renamed this file to have ctrl suffix and followed your todo hints to implement endpoint

File: integration_tests/keysend_test.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	// check that balance was reduced
+	userId := getUserIdFromToken(suite.aliceToken)
+	aliceBalance, err := suite.service.CurrentUserBalance(context.Background(), userId)
+	if err != nil {
+		fmt.Printf("Error when getting balance %v\n", err.Error())
+	}
+	assert.Equal(suite.T(), int64(aliceFundingSats)-int64(externalSatRequested), aliceBalance)
+}
+
+func (suite *KeySendTestSuite) TestKeysendPaymentNonExistentDestination() {
```
Comment (skosito) [inline integration_tests/keysend_test.go:102]: i did not add test with memo too long case, could you please clarify why it's needed @kiwiidb ?

File: lib/service/invoices.go [hunk 3]
```diff
@@ -137,6 +127,44 @@ func (svc *LndhubService) SendPaymentSync(ctx context.Context, invoice *models.I
 	return sendPaymentResponse, nil
 }
 
+func createLnRpcSendRequest(invoice *models.Invoice) (*lnrpc.SendRequest, error) {
```
Comment (skosito) [inline lib/service/invoices.go:130]: @kiwiidb i unified same code into this method to reduce duplication

File: lib/service/invoices.go [hunk 4]
```diff
@@ -235,7 +263,7 @@ func (svc *LndhubService) HandleSuccessfulPayment(ctx context.Context, invoice *
 	return err
 }
 
-func (svc *LndhubService) AddOutgoingInvoice(ctx context.Context, userID int64, paymentRequest string, decodedInvoice *lnrpc.PayReq) (*models.Invoice, error) {
+func (svc *LndhubService) AddOutgoingInvoice(ctx context.Context, userID int64, paymentRequest string, decodedInvoice *lnrpc.PayReq, keysend bool) (*models.Invoice, error) {
```
Comment (skosito) [inline lib/service/invoices.go:266]: i think we can maybe add struct that would contain both *lnrpc.PayReq and keysend, just wanted to check if it makes sense from domain perspective?

File: integration_tests/keysend_test.go (hunk 2 above)
Comment (kiwiidb) [reply integration_tests/keysend_test.go:102]: I think it's fine to leave it out then.

File: lib/service/invoices.go (hunk 4 above)
Comment (kiwiidb) [reply lib/service/invoices.go:266]: I think it makes sense, yes. Also in the future we might have bolt12 or other extensions of the PayReq. So you can implement it.

File: main.go [hunk 5]
```diff
@@ -134,6 +134,7 @@ func main() {
 	secured.GET("/checkpayment/:payment_hash", controllers.NewCheckPaymentController(svc).CheckPayment)
 	secured.GET("/balance", controllers.NewBalanceController(svc).Balance)
 	secured.GET("/getinfo", controllers.NewGetInfoController(svc).GetInfo)
+	secured.GET("/keysend", controllers.NewKeySendController(svc).KeySend)
```
Comment (kiwiidb) [inline main.go:137]: Should be POST.

**Review by kiwiidb — CHANGES_REQUESTED** [review]
- Change to `POST`
- Integrate keysend boolean and lnrpc.PayReq into single struct like you suggested.

If these things are changed, then this would LGTM.

**Comment by skosito** [comment]
> * Change to `POST`
> * Integrate keysend boolean and lnrpc.PayReq into single struct like you suggested.
> 
> If these things are changed, then this would LGTM.

@kiwiidb great catch with GET, not sure how I missed it.
I added a struct, not sure if there is a better name for it, please check.

**Comment by kiwiidb** [comment]
Better to name it "LNPayReq", we don't want extra references to LND because it could hold extra fields that are not related to LND (bolt12).

**Review by kiwiidb — APPROVED** [review]

**Comment by skosito** [comment]
> Better to name it "LNPayReq", we don't want extra references to LND because it could hold extra fields that are not related to LND (bolt12).

@kiwiidb sure, renamed it

**Pull request merged**